File: src/cli_client/managers.py
from typing import Dict, List, Set, Optional
import json
import logging

LOG_FORMAT = "\033[32m%(levelname)s\033[0m:    %(message)s"
logging.basicConfig(level=logging.INFO, format=LOG_FORMAT)
logger = logging.getLogger("managers")


class ToolManualManager:
    def __init__(self):
        # summary of all tools
        self._summaries: Dict[str, str] = {}

        # registry of the tool manuals, keyed by server name
        self._registry: Dict[str, str] = {}
        
        # The "Active Loadout"
        # These are the tools currently injected into the LLM's prompt/API call.
        self._active_tools: Set[str] = set()

    def register_summary(self, summaries: Dict[str, str]):
        """Adds a tool summary to the master summary registry."""
        for name, summary in summaries.items():
            logger.info(f"Registering tool summary '{name}'")
            self._summaries[name] = summary

    def register_tool(self, name: str, manual: str):
        """Adds a tool to the master registry (not yet visible to the LLM)."""
        logger.info(f"Registering tool manual '{name}'")
        self._registry[name] = manual
    
    def check_tool(self, name: str) -> bool:
        return name in self._registry

    def inject_tool(self, name: str) -> bool:
        """Fetch from registry and move into the active context."""
        if name in self._registry:
            self._active_tools.add(name)
            logger.info(f"Injecting tool manual '{name}'")
            return True
        return False

    def prune_tool(self, name: str):
        """Remove a specific tool from the context after a specific call."""
        if name in self._active_tools:
            self._active_tools.remove(name)
            logger.info(f"Pruning tool manual '{name}'")
    
    def flush_loadout(self):
        """Clear all tool descriptions to reset context to a 'zero-pollution' state."""
        self._active_tools.clear()
        logger.info("Flushing all tool manuals from context")

    def get_all_summaries(self) -> str:
        """Returns all summaries of the tools."""
        return '\n\n'.join([f"Tool Name: {name}\nSummary: {summary}" for name, summary in self._summaries.items()])

    def get_all_manuals(self) -> str:
        """Returns the 'tools' parameter for the LLM API call."""
        delimeter = f"\n\n{'*'*30}\n\n"
        return delimeter.join([f"Tool Name: {name}\nManual: {self._registry[name]}" for name in self._active_tools])
    
    def get_tool_info(self) -> Dict[str, str]:
        """
        Return tool summaries and active tool manuals
        """

        tool_info = {
            "tool_summaries": self.get_all_summaries(),
            "tool_manuals": self.get_all_manuals()
        }

        return tool_info

File: src_ts/cli_client/managers.ts
import { logInfo } from "../shared/logging";

export class ToolManualManager {
  private summaries: Record<string, string> = {};
  private registry: Record<string, string> = {};
  private activeTools = new Set<string>();

  registerSummary(summaries: Record<string, string>) {
    for (const [name, summary] of Object.entries(summaries)) {
      logInfo("managers", `Registering tool summary '${name}'`);
      this.summaries[name] = summary;
    }
  }

  registerTool(name: string, manual: string) {
    logInfo("managers", `Registering tool manual '${name}'`);
    this.registry[name] = manual;
  }

  checkTool(name: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.registry, name);
  }

  injectTool(name: string): boolean {
    if (this.checkTool(name)) {
      this.activeTools.add(name);
      logInfo("managers", `Injecting tool manual '${name}'`);
      return true;
    }
    return false;
  }

  pruneTool(name: string) {
    if (this.activeTools.has(name)) {
      this.activeTools.delete(name);
      logInfo("managers", `Pruning tool manual '${name}'`);
    }
  }

  flushLoadout() {
    this.activeTools.clear();
    logInfo("managers", "Flushing all tool manuals from context");
  }

  private getAllSummaries(): string {
    return Object.entries(this.summaries)
      .map(([name, summary]) => `Tool Name: ${name}\nSummary: ${summary}`)
      .join("\n\n");
  }

  private getAllManuals(): string {
    const delimiter = `\n\n${"*".repeat(30)}\n\n`;
    return Array.from(this.activeTools)
      .map((name) => `Tool Name: ${name}\nManual: ${this.registry[name]}`)
      .join(delimiter);
  }

  getToolInfo(): { tool_summaries: string; tool_manuals: string } {
    return {
      tool_summaries: this.getAllSummaries(),
      tool_manuals: this.getAllManuals(),
    };
  }
}

